tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules/SetInRuleWithArrays.scala

```diff
@@ -119,25 +119,33 @@ class SetInRuleWithArrays(rewriter: SymbStateRewriter) extends SetInRule(rewrite
       var nextState = state.updateArena(_.appendCell(BoolT1()))
       val pred = nextState.arena.topCell.toNameEx
 
-      // EqConstraints need to be generated, since missing in-relations, e.g. in sets of tuples, will lead to errors.
-      // TODO: Inlining this method is pointless. We should consider handling tuples and other structures natively in SMT.
-      // If elemCell is a function, cacheEqConstraints will generate constraints over its set of relations. This is
-      // problematic because the relations are only being carried in the arena, without SMT constrains being generated.
       elemCell.cellType match {
-        case CellTFrom(FunT1(_, _)) => ()
-        case _ =>
-          nextState = rewriter.lazyEq.cacheEqConstraints(nextState, potentialElems.map((_, elemCell)))
-      }
+        // For types that don't rely on lazy equality, we can simply query the SMT array for membership
+        case CellTFrom(_ @IntT1() | RealT1() | BoolT1() | StrT1() | ConstT1(_) | VarT1(_) | FunT1(_, _) | SetT1(_)) =>
+          val eql = tla.eql(pred, tla.apalacheSelectInSet(elemCell.toNameEx, setCell.toNameEx))
+          rewriter.solverContext.assertGroundExpr(eql)
+          nextState.setRex(pred)
 
-      def inAndEq(elem: ArenaCell) = {
-        simplifier
-          .simplifyShallow(tla.and(tla.apalacheSelectInSet(elem.toNameEx, setCell.toNameEx),
-                  tla.eql(elem.toNameEx, elemCell.toNameEx)))
+        // If the element's type does not support SMT equality, we rely on lazy equality
+        case _ =>
```

Comment:
I'm wondering if it would help with clarify the intent of the code here to introduce a predicate and do something like:

```scala
if requiresLazyEquality(elemCell.cellType) {
  // ...
} else {
  // ...
}
```

I think this captures most of what is currently in these two comments on the `case` branches?